### DISCUSSION TITLE
How to change every single character's color in one text?

### DISCUSSION THREAD

**Discussion by Capricius** [typst/typst#3947] [Q&A]
Is this possible in typst? such as a word "selengensis", can we only change a few letters in it. Or like a word "حالف", can we change the middle letter's color?

**Comment by Coekjan** [comment]
Like this?

```typst
ab#text(fill: red, "cde")fg
```

![image](https://github.com/typst/typst/assets/69834864/e51bf22d-d345-43b4-844b-01fb9dd30c67)

**Reply by Capricius** [reply]
Yes, But now I realized that some languages not supported. Especially the arabic ligature will be cut due to color-change.
![图片1](https://github.com/typst/typst/assets/71160642/7b421875-cd8b-45c0-bc70-e28b75f69889)

**Reply by laurmaedje** [reply]
If you change the color mid-ligature, there isn't much we can do, I think. Note that just `set text(blue)` should suffice, the `show ch` part scopes the set rule to just the specific letters, but at that point you have ensured that only this letter exists. Alternatively, you can remove the whole loop and write `show "a": set text(blue)`.

**Reply by Capricius** [reply]
Thanks, Laurmaedje !

**Reply by Enter-tainer** [reply]
> If you change the color mid-ligature, there isn't much we can do

addtional note explaining why it is not possible. when typesetting, a ligature is a single glyph and we cannot tell where to divide the parts, thus to color them individually.

**Reply by Capricius** [reply]
But how Microsoft Word & Powerpoint to achieve like this:
![fig1](https://github.com/typst/typst/assets/71160642/9f8d1021-12aa-41df-94f8-7876bd30e3fd)

**Reply by axelkar** [reply]
I remembered [an interesting example from Pango](https://docs.gtk.org/Pango/pango_markup.html#pango-markup), but it doesn't seem like it can color ligatures with multiple colors:
```console
$ pango-view --markup --font="sans 120" -t '<span foreground="red">f</span><span foreground="blue">i</span>'
# shows f and i in blue
```
Not sure how Arabic text works there
